pkg/controller/zone_api.go

```diff
@@ -178,3 +181,92 @@ func (s *ZoneApi) getZoneByID(ctx context.Context, id string) (*edgeproto.Zone,
 	}
 	return zone, nil
 }
+
+func (s *ZoneApi) ShowZoneGPUs(filter *edgeproto.Zone, cb edgeproto.ZoneApi_ShowZoneGPUsServer) error {
+	ctx := cb.Context()
+	// get all matching zones
+	zones := []edgeproto.ZoneKey{}
+	err := s.cache.Show(filter, func(zone *edgeproto.Zone) error {
+		zones = append(zones, zone.Key)
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	sort.Slice(zones, func(i, j int) bool {
+		return zones[i].GetKeyString() < zones[j].GetKeyString()
+	})
+	for _, zkey := range zones {
+		zoneGPUs, err := s.getZoneGPUs(ctx, &zkey)
+		if err != nil {
+			log.SpanLog(ctx, log.DebugLevelApi, "failed to get GPUs for zone", "zone", zkey, "err", err)
+			continue
+		}
+		if len(zoneGPUs.Gpus) == 0 {
+			continue
+		}
+		err = cb.Send(zoneGPUs)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *ZoneApi) getZoneGPUs(ctx context.Context, zkey *edgeproto.ZoneKey) (*edgeproto.ZoneGPUs, error) {
+	// get all cloudlets for zone
+	ckeys := []edgeproto.CloudletKey{}
+	err := s.all.cloudletApi.cache.Show(&edgeproto.Cloudlet{}, func(cloudlet *edgeproto.Cloudlet) error {
+		if cloudlet.Key.Organization != zkey.Organization {
+			return nil
+		}
+		if cloudlet.Zone != zkey.Name {
+			return nil
+		}
+		ckeys = append(ckeys, cloudlet.Key)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	usedVals := resspec.ResValMap{}
+	limits := ResLimitMap{}
+	gpuInfos := map[string]*edgeproto.GPUResource{}
+
+	for _, ckey := range ckeys {
+		if err := s.all.cloudletApi.addGPUsUsage(ctx, &ckey, usedVals, limits, gpuInfos); err != nil {
+			return nil, err
+		}
+	}
+	gpus := []*edgeproto.GPUResource{}
+	for _, gpu := range gpuInfos {
+		resKey := edgeproto.BuildResKey(cloudcommon.ResourceTypeGPU, gpu.ModelId)
+		max := 0
+		if limit, ok := limits[resKey]; ok {
+			if limit.QuotaMaxValue > 0 {
+				max = int(limit.QuotaMaxValue)
+			} else {
+				max = int(limit.InfraMaxValue)
+			}
+		}
+		used := 0
+		if res, ok := usedVals[resKey]; ok {
+			used = int(res.Value.Whole)
+		}
+		// set gpu count to 1 if there are gpus available
+		if max == 0 || max > used {
```

Comment:
If max is 0, do we just assume that number of gpus is unlimited?

Reply:
Yup, max is 0 is unlimited. It's the case where the infra is not reporting what the upper limit is.